gst/nnstreamer/tensor_query/tensor_query_server.c

```diff
@@ -31,11 +31,33 @@ static GHashTable *_qs_table = NULL;
 static void init_queryserver (void) __attribute__((constructor));
 static void fini_queryserver (void) __attribute__((destructor));
 
+/**
+ * @brief Internal function to release query server data.
+ */
+static void
+_release_server_data (gpointer data)
+{
+  GstTensorQueryServer *_data = (GstTensorQueryServer *) data;
+
+  if (!_data)
+    return;
+
+  if (_data->edge_h) {
+    nns_edge_release_handle (_data->edge_h);
+    _data->edge_h = NULL;
+  }
+
+  g_cond_clear (&_data->cond);
+  g_mutex_clear (&_data->lock);
+  g_free (_data->id);
+  g_free (_data);
```

Comment:
What if someone else is holding ```_data->lock``` while this function is called?

Reply:
updated

Reply:
It would be safer to do ```g_mutex_unlock``` after ```g_free (_data)```. You can add ```g_mutex *tmp = &_data->lock;``` and control the mutex with ```tmp```.

Reply:
updated

Reply:
uh...

```
g_cond_clear ()
void                g_cond_clear                        (GCond *cond);
Frees the resources allocated to a GCond with g_cond_init().

This function should not be used with a GCond that has been statically allocated.

Calling g_cond_clear() for a GCond on which threads are blocking leads to undefined behaviour.

cond :

an initialised GCond
```


looks like a trouble for g_cond_clear() here...

Reply:
rollbacked, _release_server_data() will be called with table-lock.